drivers/lte_link_control/lte_lc.c

```diff
@@ -166,24 +174,51 @@ static int w_lte_lc_init(void)
 
 static int w_lte_lc_connect(void)
 {
-	if (at_cmd_write(network_mode, NULL, 0, NULL) != 0) {
-		return -EIO;
-	}
+	int err;
+	char *current_network_mode = (char *)nw_mode_preferred;
+	bool retry;
 
 	k_sem_init(&link, 0, 1);
-	at_params_list_init(&params, 10);
 	at_cmd_set_notification_handler(at_handler);
+	at_params_list_init(&params, 10);
 
-	if (at_cmd_write(normal, NULL, 0, NULL) != 0) {
-		return -EIO;
-	}
+	do {
+		retry = false;
 
-	k_sem_take(&link, K_FOREVER);
+		LOG_DBG("Network mode: %s", log_strdup(current_network_mode));
+
+		if (at_cmd_write(current_network_mode, NULL, 0, NULL) != 0) {
+			return -EIO;
+		}
+
+		if (at_cmd_write(normal, NULL, 0, NULL) != 0) {
+			return -EIO;
+		}
+
+		err = k_sem_take(&link, K_SECONDS(CONFIG_LTE_NETWORK_TIMEOUT));
+		if (err == -EAGAIN) {
+			LOG_INF("Network connection attempt timed out");
+
+			if (IS_ENABLED(CONFIG_LTE_NETWORK_USE_FALLBACK) &&
+			    (current_network_mode == nw_mode_preferred)) {
+				current_network_mode = (char *)nw_mode_fallback;
+				retry = true;
+
+				if (at_cmd_write(offline, NULL, 0, NULL) != 0) {
+					return -EIO;
```

Comment:
Do you not need to call `at_params_list_free` before returning, or remove the notification handler?

Reply:
Good catch. Didn't notice it coming in after rebase. Will fix it.

Reply:
Fixed, (force-push diff after your approval here, @rlubos : https://github.com/NordicPlayground/fw-nrfconnect-nrf/compare/87f0658745a9d8c5876a405944357a1cf6348742..a0d2740e9c7b0b611731ff3a55c55da9a641cea4)

Reply:
Thanks, still LGTM, nice catch.